PD client should handle errors properly

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->

v4.0.15

### What operating system and CPU are you using?
<!-- If you're using Linux, you can run `cat /proc/cpuinfo` -->

doesn't matter

### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->

Once PD client meets an error, it will reconnect and retry the request even if the request is invalid. The disconnection may result in other requests failure, e.g., region heartbeat. I saw lots of logs below. I suspect the disconnection makes region heartbeat fail and PD contains stale region information for a long time.
```
[2022/04/06 19:03:05.883 +08:00] [ERROR] [util.rs:348] ["request failed, retry"] [err_code=KV:Unknown] [err="Grpc(RpcFailure(RpcStatus { status: 2-UNKNOWN, details: Some(\"region not found, request region: id:56125222 start_key:\\\"7480000000014B7AFF3500000000000000F8\\\" region_epoch:<conf_ver:288104 version:4789963 > peers:<id:175253496 store_id:161941895 > peers:<id:175254079 store_id:161941897 > peers:<id:175443528 store_id:155883000 >\") }))"]
[2022/04/06 19:03:05.883 +08:00] [INFO] [util.rs:470] ["connecting to PD endpoint"] [endpoints=http://172.29.5.31:2379]
[2022/04/06 19:03:05.884 +08:00] [INFO] [<unknown>] ["New connected subchannel at 0x7fa29ebb7e90 for subchannel 0x7f99d359f2c0"]
[2022/04/06 19:03:05.885 +08:00] [INFO] [util.rs:470] ["connecting to PD endpoint"] [endpoints=http://172.29.5.33:2379]
[2022/04/06 19:03:05.886 +08:00] [INFO] [util.rs:533] ["connected to PD leader"] [endpoints=http://172.29.5.33:2379]
[2022/04/06 19:03:05.886 +08:00] [INFO] [util.rs:231] ["heartbeat sender and receiver are stale, refreshing ..."]
[2022/04/06 19:03:05.997 +08:00] [WARN] [util.rs:249] ["updating PD client done"] [spend=114.438081ms]
```
### What did you expect?

PD client doesn't reconnect and retry invalid requests.

### What did happened?

PD client does reconnect and retry.

--- a/components/pd_client/src/errors.rs
+++ b/components/pd_client/src/errors.rs
@@ -30,8 +30,9 @@ pub type Result<T> = result::Result<T, Error>;
 impl Error {
     pub fn retryable(&self) -> bool {
         match self {
-            Error::Grpc(_) | Error::Other(_) | Error::ClusterNotBootstrapped(_) => true,
-            Error::RegionNotFound(_)
+            Error::Grpc(_) | Error::ClusterNotBootstrapped(_) => true,
+            Error::Other(_)
+            | Error::RegionNotFound(_)
             | Error::StoreTombstone(_)
             | Error::GlobalConfigNotFound(_)
             | Error::ClusterBootstrapped(_)

--- a/components/test_pd/src/mocker/retry.rs
+++ b/components/test_pd/src/mocker/retry.rs
@@ -91,7 +91,7 @@ impl PdMocker for NotRetry {
                 "[NotRetry] get_region_by_id returns Ok(_) with header has IncompatibleVersion error"
             );
             let mut err = Error::default();
-            err.set_type(ErrorType::IncompatibleVersion);
+            err.set_type(ErrorType::RegionNotFound);
             let mut resp = GetRegionResponse::default();
             resp.mut_header().set_error(err);
             Some(Ok(resp))
@@ -107,7 +107,7 @@ impl PdMocker for NotRetry {
                 "[NotRetry] get_region_by_id returns Ok(_) with header has IncompatibleVersion error"
             );
             let mut err = Error::default();
-            err.set_type(ErrorType::IncompatibleVersion);
+            err.set_type(ErrorType::Unknown);
             let mut resp = GetStoreResponse::default();
             resp.mut_header().set_error(err);
             Some(Ok(resp))

--- a/components/test_pd/src/mocker/retry.rs
+++ b/components/test_pd/src/mocker/retry.rs
@@ -88,7 +88,7 @@ impl PdMocker for NotRetry {
     fn get_region_by_id(&self, _: &GetRegionByIdRequest) -> Option<Result<GetRegionResponse>> {
         if !self.is_visited.swap(true, Ordering::Relaxed) {
             info!(
-                "[NotRetry] get_region_by_id returns Ok(_) with header has IncompatibleVersion error"
+                "[NotRetry] get_region_by_id returns Ok(_) with header has RegionNotFound error"
             );
             let mut err = Error::default();
             err.set_type(ErrorType::RegionNotFound);
@@ -104,7 +104,7 @@ impl PdMocker for NotRetry {
     fn get_store(&self, _: &GetStoreRequest) -> Option<Result<GetStoreResponse>> {
         if !self.is_visited.swap(true, Ordering::Relaxed) {
             info!(
-                "[NotRetry] get_region_by_id returns Ok(_) with header has IncompatibleVersion error"
+                "[NotRetry] get_region_by_id returns Ok(_) with header has Unknown error"
             );
             let mut err = Error::default();
             err.set_type(ErrorType::Unknown);

--- a/components/test_pd/src/mocker/retry.rs
+++ b/components/test_pd/src/mocker/retry.rs
@@ -87,9 +87,7 @@ impl Default for NotRetry {
 impl PdMocker for NotRetry {
     fn get_region_by_id(&self, _: &GetRegionByIdRequest) -> Option<Result<GetRegionResponse>> {
         if !self.is_visited.swap(true, Ordering::Relaxed) {
-            info!(
-                "[NotRetry] get_region_by_id returns Ok(_) with header has RegionNotFound error"
-            );
+            info!("[NotRetry] get_region_by_id returns Ok(_) with header has RegionNotFound error");
             let mut err = Error::default();
             err.set_type(ErrorType::RegionNotFound);
             let mut resp = GetRegionResponse::default();
@@ -103,9 +101,7 @@ impl PdMocker for NotRetry {
 
     fn get_store(&self, _: &GetStoreRequest) -> Option<Result<GetStoreResponse>> {
         if !self.is_visited.swap(true, Ordering::Relaxed) {
-            info!(
-                "[NotRetry] get_region_by_id returns Ok(_) with header has Unknown error"
-            );
+            info!("[NotRetry] get_region_by_id returns Ok(_) with header has Unknown error");
             let mut err = Error::default();
             err.set_type(ErrorType::Unknown);
             let mut resp = GetStoreResponse::default();

